Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/type/common/TestCertificateRetrievalScheduler.kt

```diff
@@ -0,0 +1,116 @@
+
+@Singleton
+class TestCertificateRetrievalScheduler @Inject constructor(
+    @AppScope private val appScope: CoroutineScope,
+    private val workManager: WorkManager,
+    private val certificateRepo: TestCertificateRepository,
+    private val testRepo: CoronaTestRepository,
+    private val foregroundState: ForegroundState,
+) : ResultScheduler(
+    workManager = workManager
+) {
+    private val processedNewCerts = mutableSetOf<String>()
+
+    private val creationTrigger = testRepo.coronaTests
+        .map { tests ->
+            tests
+                .filter { it.isDccSupportedByPoc } // Only those that support it
+                .filter { it.isNegative } // Certs only to proof negative state
+                .filter { it.isDccConsentGiven && !it.isDccDataSetCreated } // Consent and doesn't exist already?
+        }
+        .distinctUntilChanged()
+
+    private val refreshTrigger = combine(
+        certificateRepo.certificates,
+        foregroundState.isInForeground,
+    ) { certificates, isForeground ->
+
+        val hasNewCert = certificates.any {
+            val isNew = processedNewCerts.contains(it.identifier)
+            if (isNew) processedNewCerts.add(it.identifier)
+            isNew
+        }
+
+        val hasWorkToDo = certificates.any { it.isPending }
+        Timber.tag(TAG).v("shouldPollDcc? hasNewCert=$hasNewCert, hasWorkTodo=$hasWorkToDo, foreground=$isForeground")
+        (isForeground || hasNewCert) && hasWorkToDo
+    }
+        .distinctUntilChanged()
+
+    fun setup() {
+        Timber.tag(TAG).i("setup() - TestCertificateRetrievalScheduler")
+
+        // Create a certificate entry for each viable test that has none
+        creationTrigger
+            .onEach { testsWithoutCert ->
+                Timber.tag(TAG).d("State change: testsWithoutCert=$testsWithoutCert")
+                testsWithoutCert.forEach { test ->
+                    val cert = certificateRepo.requestCertificate(test)
+                    Timber.tag(TAG).v("Certificate was created: %s", cert)
+                    testRepo.markDccAsCreated(test.identifier, created = true)
+                }
+            }
+            .launchIn(appScope)
+
+        // For each change to the set of existing certificates, check if we need to refresh/load data
+        refreshTrigger
+            .onEach { checkCerts ->
+                Timber.tag(TAG).d("State change: checkCerts=$checkCerts")
+                if (checkCerts) scheduleWorker()
+            }
+            .launchIn(appScope)
+    }
+
+    internal suspend fun scheduleWorker() {
+        Timber.tag(TAG).i("scheduleWorker()")
+
+        if (isScheduled(WORKER_ID)) {
+            Timber.tag(TAG).d("Worker already queued, skipping requeue.")
+        }
+
+        Timber.tag(TAG).d("enqueueUniqueWork PCR_TESTRESULT_WORKER_UNIQUEUNAME")
+        workManager.enqueueUniqueWork(
+            WORKER_ID,
+            ExistingWorkPolicy.KEEP,
+            buildWorkerRequest()
+        )
+    }
```

Comment:
Test certificate retrieval happens via worker, if not riggered through the UI.
We may receive a negative test result while the app is in the background, and to not be killed off by the system, before we can get a certificate, we need to run the `TestCertificateRepo.refresh()` within a worker.